### PR TITLE
S8 Find User screen implementation & logic

### PR DESCRIPTION
**Pull request by albertfares**
# Find user screen implementation & logic

## Overview

This PR introduces the "Add Friend" screen to the GoMeet app. The following functionalities have been implemented:

1. **User List Fetching and Display**: Fetch all users from Firestore and display them in a scrollable list, excluding the current user.
2. **Search Functionality**: Implement a search bar to filter the user list based on a search query.
3. **Follow/Unfollow Users**: Allow users to follow or unfollow other users with immediate state updates.
4. **Profile Image Display**: Display user profile images in the user list.
5. **Navigation to User Profiles**: Navigate to a user's profile when clicking on a user item.
6. **UI Enhancements**: Improve the overall user experience with responsive and interactive UI components.

## Changes

- **Add Friend Screen**: Created a new screen (`AddFriend`) to allow users to find and add friends.
  - Utilizes `Scaffold` for consistent screen layout.
  - Includes a top app bar with a back navigation button.
  - Features a search bar at the top to filter the list of users.
  
- **User List**: Fetches the list of users from Firestore and displays it in a `LazyColumn`, excluding the current user.
  - Users are displayed with their profile picture, username, and full name.
  - Fetches the list of users asynchronously and displays a loading indicator (`CircularProgressIndicator`) while fetching.
  
- **Search Bar**: Added a search bar (`GoMeetSearchBar`) to filter users based on their username.
  - Updates the displayed list of users in real-time as the search query changes.

- **Follow/Unfollow Functionality**: Implemented follow and unfollow functionality.
  - The follow button changes to "Unfollow" if the user is already being followed and vice versa.
  - Uses immediate state updates to reflect changes in the UI without delay.
  - Follow and unfollow actions are handled asynchronously.

- **Profile Images**: Displays profile images for each user in the list using the `ProfileImage` composable.
  - Fetches profile images from Firestore.
  - Displays a default image if no profile picture is available.

- **Navigation**: Navigates to the selected user's profile when a user item is clicked.
  - Uses the `NavigationActions` object to handle navigation.
  - Navigates to `Route.OTHERS_PROFILE`, replacing the placeholder with the selected user's UID.

- **UI Enhancements**:
  - Improved the visual layout with responsive spacing and padding.
  - Ensured consistency in the design by using `MaterialTheme` for colors and typography.
  - Added click animations and visual feedback for better user interaction.

## Screenshot

Screenshot 2024-05-15 at 18 03 48

These enhancements provide a comprehensive and user-friendly way for users to find and connect with others within the GoMeet app.

### CODE REVIEW

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            users
+          } else {
+            users.filter { it.username.contains(query.value, ignoreCase = true) }
+          }
+        }
+      }
+
+  Scaffold(
+      topBar = {
+        Column {
```
Comment (chorfiyoussef) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:81]: This Column is useless

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+              modifier = Modifier.testTag("TopBar"),
+              backgroundColor = MaterialTheme.colorScheme.background,
+              elevation = 0.dp,
+              title = {
+                // Empty title since we're placing our own components
+              },
+              navigationIcon = {
+                IconButton(onClick = { nav.goBack() }) {
+                  Icon(
+                      imageVector = Icons.AutoMirrored.Filled.ArrowBack,
```
Comment (chorfiyoussef) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:92]: Actual Icon used for a "Go Back" Button is `Icons.AutoMirrored.Filled.KeyboardArrowLeft`.
I think we should stay consistent on that
Capture d’écran 2024-05-15 à 18 30 40

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        }
+        Button(
+            colors =
+                ButtonDefaults.buttonColors(
+                    if (isFollowing) MaterialTheme.colorScheme.outlineVariant else Color.White,
+                    if (isFollowing) MaterialTheme.colorScheme.onSecondary
+                    else MaterialTheme.colorScheme.tertiary),
+            onClick = { onFollowButtonClick(user.uid, isFollowing) }) {
+              Text(text = if (isFollowing) "Unfollow" else "Follow")
+            }
```
Comment (chorfiyoussef) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:216]: Button shape, size and colors doesn't match the buttons in Followers / Following screens.

For reference, here is the Button settings in thos screens :
Button(
                    shape = RoundedCornerShape(10.dp),
                    onClick = {
                      userViewModel.unfollow(user.uid)
                      isFollowing = false
                    },
                    modifier =
                        Modifier.padding(start = 15.dp).width(110.dp).testTag("UnfollowButton"),
                    colors =
                        ButtonColors(
                            containerColor = MaterialTheme.colorScheme.primaryContainer,
                            contentColor = Color.Black,
                            disabledContentColor = Color.Transparent,
                            disabledContainerColor = Color.Transparent)) {
                      Text(text =  insert your text, style = MaterialTheme.typography.labelLarge)
                    }
                    
Here is a picture demonstrating these differences :
Capture d’écran 2024-05-15 à 18 39 02
Capture d’écran 2024-05-15 à 18 39 21

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+          Row(
+              verticalAlignment = Alignment.CenterVertically,
+              modifier = Modifier.padding(start = screenWidth / 15, top = 0.dp)) {
+                Text(
+                    text = "Find User",
+                    style =
+                        MaterialTheme.typography.headlineMedium.copy(
+                            fontWeight = FontWeight.SemiBold))
+              }
```
Comment (chorfiyoussef) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:106]: Text might look better : smaller and on the same horizontal height as the goBack Button.
What are your thoughts on this change ?

**Review by chorfiyoussef — CHANGES_REQUESTED** [review]
Code is clean and UI looks good overall, the search bar works well.

This functionality enhances the App, as it is now possible to look for a user in order to follow him. However, please ensure the code remains consistent with the rest of the app's styling and functionality, particularly concerning icons and button's shape, size, and colors. 

Also, please check my suggestions.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt (hunk 1 above)
Comment (albertfares) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:81]: Removing it makes the Screen title and go back button overlap.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt (hunk 2 above)
Comment (albertfares) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:92]: The group thinks that the `Icons.AutoMirrored.Filled.ArrowBack` fits better than the `KeyboardArrowLeft`

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt (hunk 3 above)
Comment (albertfares) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:216]: Fixed. Thanks for the feedback!

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt (hunk 4 above)
Comment (albertfares) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:106]: It does look better, thanks!

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt (hunk 2 above)
Comment (chorfiyoussef) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/profile/AddFriend.kt:93]: Ok ! I will change screens that does not use this button!

**Review by chorfiyoussef — APPROVED** [review]
This screen is nicely done and I can assure that it is consistent with the UI of the rest of the App. Thank you Albert!

**Pull request merged**